akka-cluster/src/main/scala/akka/cluster/sbr/DowningStrategy.scala

```diff
@@ -281,30 +282,35 @@ import akka.coordination.lease.scaladsl.Lease
     }
   }
 
-  private def additionalNodesToDownWhenIndirectlyConnected: Set[UniqueAddress] = {
+  private def additionalNodesToDownWhenIndirectlyConnected(downable: Set[UniqueAddress]): Set[UniqueAddress] = {
     if (unreachableButNotIndirectlyConnected.isEmpty)
       Set.empty
     else {
       val originalUnreachable = _unreachable
       val originalReachability = _reachability
+
       try {
         val intersectionOfObserversAndSubjects = indirectlyConnectedFromIntersectionOfObserversAndSubjects
         val haveSeenCurrentGossip = indirectlyConnectedFromSeenCurrentGossip
-        // remove records between the indirectly connected
-        _reachability = reachability.filterRecords(
-          r =>
-            !((intersectionOfObserversAndSubjects(r.observer) && intersectionOfObserversAndSubjects(r.subject)) ||
-            (haveSeenCurrentGossip(r.observer) && haveSeenCurrentGossip(r.subject))))
+        _reachability = reachability
+          .filterRecords { r =>
+            // we only retain records for addresses that are still downable
+            downable.contains(r.observer) && downable.contains(r.subject) &&
```

Comment:
If we have a split with indirectly connected nodes during a rolling upgrade, we may have records with nodes that are already leaving or downed. 
When that happens, the filter below won't be able to remove the indirectly connected nodes causing the next decision to switch to `DownAll` through the `IllegalStateException`.

So the fix is to remove all records that are not downable.